src/action-ros-ci.ts

```diff
@@ -225,16 +226,66 @@ async function run() {
 		await io.rmRF(rosWorkspaceDir);
 
 		// Checkout ROS 2 from source and install ROS 2 system dependencies
-		await io.mkdirP(rosWorkspaceDir + "/src");
+		await io.mkdirP(path.join(rosWorkspaceDir, "src"));
+
+		// Download repos files into a specific directory
+		const reposFilesDir = path.join(rosWorkspaceDir, "repos-files");
+		await io.mkdirP(reposFilesDir);
+		const curlFlags = curlFlagsArray.join(" ");
+		const reposFiles: string[] = [];
+		for (const [
+			index,
+			vcsRepoFileUrl,
+		] of vcsRepoFileUrlListResolved.entries()) {
+			const reposFile = `${index}.repos`;
+			await execBashCommand(
+				`curl ${curlFlags} '${vcsRepoFileUrl}' --output ${reposFile}`,
+				undefined,
+				{ cwd: reposFilesDir }
+			);
+			reposFiles.push(path.join(reposFilesDir, reposFile));
+		}
 
 		const options = {
 			cwd: rosWorkspaceDir,
 		};
 
-		const curlFlags = curlFlagsArray.join(" ");
-		for (const vcsRepoFileUrl of vcsRepoFileUrlListResolved) {
+		// Process repos files and apply PR dependency changes
+		const prDependencies = dep.getPrDependencies(github.context.payload);
+		if (prDependencies.length === 0) {
+			await core.group("No PR dependencies", () => Promise.resolve());
+		} else {
+			await core.group("PR dependencies", () => {
+				for (const dependency of prDependencies) {
+					core.info("\t" + dep.getFullUrlFromDependency(dependency));
+				}
+				return Promise.resolve();
+			});
+			// Configure git to fetch/checkout PR & MR refs
+			await execBashCommand(
+				"git config --global --add remote.origin.fetch '+refs/pull/*:refs/remotes/origin/pull/*' && " +
+					"git config --global --add remote.origin.fetch '+refs/merge-requests/*/head:refs/remotes/origin/merge-requests/*'",
+				undefined,
+				options
+			);
+			// Modify repos files
+			const extraReposFilePath = dep.replaceDependencyVersions(
```

Comment:
High level - this logic would look a lot cleaner to me if we avoided mutating the input files entirely, in favor of a simple overlay scheme - pseudocode to show what i mean:

```
const prDependencies = dep.getPrDependencies(github.context.payload);
const dependenciesRepoPath = tempfile();
dep.writeDependenciesToReposFile(prDependencies, dependenciesRepoPath);

for (const vcsRepoFileUrl of vcsRepoFileUrlListResolved) {
  execBash(`curl ${vcsRepoFileUrl} | vcs import --force --recursive src/`)
}
execBash(`vcs import --force --recursive src/ < dependenciesRepoPath`);
```

The `--force` option already forces an overwrite of existing repositories in the path, so a modification doesn't make it any more logically correct - and this way I think we could get rid of at least a hundred lines of matching logic in dependencies.ts

Does this sound good, or am I missing some crucial point?

Reply:
I don't think that works, because `vcs import --force [--recursive]` only overwrites repos (i.e. checks out another branch in our case) if directory paths match.

Here's a counter-example. Let's say we have this `.repos` file as a "base" file (i.e. element in `vcsRepoFileUrlListResolved`).

```yaml
# original.repos, e.g. ros2.repos
repositories:
  ros-tracing/ros2_tracing:
    type: git
    url: 'https://gitlab.com/ros-tracing/ros2_tracing.git'
    version: master
```

and then this file gets generated from declared dependencies (`dependenciesRepoPath`):

```yaml
repositories:
  ros2_tracing:
    type: git
    url: 'https://gitlab.com/ros-tracing/ros2_tracing.git'
    version: foxy
```

Result:

```sh
$ vcs import --force --recursive test < original.repos 
=== test/ros-tracing/ros2_tracing (git) ===
Cloning into '.'...

$ ll test 
total 0
drwxr-xr-x  3 chris  staff   96 28 Jan 17:09 .
drwxr-xr-x  8 chris  staff  256 28 Jan 17:09 ..
drwxr-xr-x  3 chris  staff   96 28 Jan 17:09 ros-tracing
$ vcs import --force --recursive test < dependencies.repos 
=== test/ros2_tracing (git) ===
Cloning into '.'...

$ ll test 
total 0
drwxr-xr-x   4 chris  staff  128 28 Jan 17:09 .
drwxr-xr-x   8 chris  staff  256 28 Jan 17:09 ..
drwxr-xr-x   3 chris  staff   96 28 Jan 17:09 ros-tracing
drwxr-xr-x  20 chris  staff  640 28 Jan 17:09 ros2_tracing
```

We just get a second repo, because the directory paths don't match. This is because we only provide URLs for dependencies; we don't provide a directory path to be used by `dep.writeDependenciesToReposFile()` in your example. We instead have to compare repo URLs, which is what I did. And I think providing anything more than a branch/MR/PR URL in PR descriptions would be rather cumbersome.

Now it's your turn to let me know if I'm missing some crucial point 😆

Reply:
hmm.. in practice, the `ros2.repos` base file has the repositories structured exclusively as

```
org/repo:
  type: git
  url: https://github.com/org/repo
```

We could easily force our output `.repos` file to match that convention. I think you're right that we can't depend on conventions for correctness though, that's going to lead to problems later - the desired directory structure isn't always `src/org/repo`.

This conversation however makes me realize a much larger problem with this feature as written - it won't work when PRs are coming from forks (unless I'm missing some crucial point) - the URL of the source repository is different than that of the target repository. This will mean that anybody without write access to the source project can't use this feature. I have solved this in practice for my script that kicks of Jenkins jobs by actually using github APIs to understand pull requests - here is the function, it should give the idea

``` python
def create_ci_gist(
    github_instance: Github,
    pulls: List[github.PullRequest.PullRequest],
    target_release: str
) -> github.Gist.Gist:
    """Create gist for the list of pull requests."""
    logger.info("Creating ros2.repos Gist for PRs")
    master_repos = fetch_repos(target_release)
    for github_pr in pulls:
        pr_ref = github_pr.head.ref
        pr_repo = github_pr.head.repo.full_name
        base_repo = github_pr.base.repo.full_name

        # Remove the existing repository from the list
        repo_entry = master_repos.pop(base_repo, None)
        if repo_entry:
            if repo_entry['type'] != 'git':
                panic('Chosen repository is not git-based')
        else:
            logger.warn(
                f'PR Repository "{pr_repo}" not found in ros2.repos, be aware that this is not '
                'part of the default build set and is not guaranteed to work.')

        # Add the same package from the PR's repository to the list
        master_repos[pr_repo] = {
            'type': 'git',
            'url': 'https://github.com/{}.git'.format(pr_repo),
            'version': pr_ref
        }

    yaml_out = yaml.dump({'repositories': master_repos}, default_flow_style=False)

    input_file = InputFileContent(content=yaml_out)
    gist = github_instance.get_user().create_gist(
        public=True,
        files={'ros2.repos': input_file},
        description='CI input for PR {}'.format(github_pr.url))
    return gist
```

the above as-is wouldn't help with gitlab or branches - but with a little extra effort, i think this could be written using the Javascript Octokit(github) and something like https://www.npmjs.com/package/gitlab.

I think if we took the above approach, we could either modify input `.repos` files, or create an overlay with the same paths for a given target repository - it seems this would be better than the complex regex logic, which is something I'm really trying to avoid having to maintain :)

Reply:
Oh - it seems like this might know how to handle forked versions - I'm just not understanding how that part works yet

Reply:
> But - I think you're right that we can't depend on conventions for correctness though, that's going to lead to problems later - the desired directory structure isn't always src/org/repo.

yeah!

> This conversation however makes me realize a much larger problem with this feature as written - it won't work when PRs are coming from forks (unless I'm missing some crucial point) - the URL of the source repository is different than that of the target repository. This will mean that anybody without write access to the source project can't use this feature.

> Oh - it seems like this might know how to handle forked versions - I'm just not understanding how that part works yet

I had to double check because I didn't remember 😆, but yeah it doesn't really understand or support forks. It only compares URLs.

I thought that perhaps this would still be useful (i.e. better than nothing), but if you think we should support forks "from day 1," I understand.

I'll work on adding fork awareness using the GitHub & GitLab APIs.